drivers/i2c/dw_i2c_master.c

```diff
@@ -1,3 +1,104 @@
 #include <target/i2c.h>
 #include <driver/dw_i2c.h>
 
+#define DW_I2C_DEBUG
+#ifdef DW_I2C_DEBUG
+#include <target/console.h>
+#endif
+
+static caddr_t base = DW_I2C_BASE(0);
+static i2c_addr_t slave_addr = 0;
+
+/* Initialize controller as master device */
+void i2c_hw_ctrl_init(void)
+{
+#ifdef DW_I2C_DEBUG
+	con_printf("Debug: Enter %s", __func__);
+#endif
+	__raw_writel(0, base + IC_ENABLE);
+	__raw_writel(IC_CON_SD /* IC_SLAVE_DISABLE */
+				 | IC_CON_RE /* IC_RESTRART_EN */
+				 /* IC_10BITADDR_MASTER = 0 */
+				 /* IC_10BITADDR_SLAVE = 0 */
+				 | IC_CON_SPD_FS /* SEEP = Fast */
+				 | IC_CON_MM /* IC_MASTER_MODE */
+				 , base + IC_CON);
+	__raw_writel(IC_RX_TL, base + IC_RX_TL);
+	__raw_writel(IC_TX_TL, base + IC_TX_TL);
+	__raw_writel(IC_STOP_DET, base + IC_INTR_MASK);
+	__raw_writel(1, base + IC_ENABLE);
+}
+
+/* Set address when this controller acts as a slave device */
+void i2c_hw_set_address(i2c_addr_t addr, boolean call)
+{
+#ifdef DW_I2C_DEBUG
+	con_printf("Debug: Enter %s", __func__);
+#endif
+	__raw_writel(0, base + IC_ENABLE);
+	__raw_writel((uint32_t)addr, base + IC_SAR);
+	__raw_writel(1, base + IC_ENABLE);
+}
+
+/* XXX */
+#ifndef CLK_CPU
+#define CLK_CPU 2000000000
+#endif
```

Comment:
This seems unbelievably fast. Please confirm with your hardware spec.

Reply:
Default is 125000 (kHz). Fixed.